magma/types/decoupled.py

```diff
@@ -0,0 +1,167 @@
+import magma as m
+from magma.tuple import Product, ProductKind
+
+
+from magma.bit import Bit
+from magma.t import Type, In, Out
+from magma.tuple import Product, ProductKind
+
+
+class ReadyValidException(Exception):
+    pass
+
+
+class ReadyValidKind(ProductKind):
+    def __getitem__(cls, T: Type):
+        if not issubclass(T, Type):
+            raise TypeError(
+                f"ReadyValid[T] expected T to be a subclass of m.Type not {T}"
+            )
+        fields = {"valid": Out(Bit), "data": Out(T), "ready": In(Bit)}
+        return type(f"ReadyValid[{T}]", (ReadyValid, ), fields)
+
+
+class ReadyValid(Product, metaclass=ReadyValidKind):
+    """
+    An I/O Bundle containing 'valid' and 'ready' signals that handshake
+    the transfer of data stored in the 'data' subfield.
+
+    The base protocol implied by the directionality is that
+    the producer uses the interface as-is (outputs bits)
+    while the consumer uses the flipped interface (inputs bits).
+
+    For readability, we provide these helpers to avoid the confusion related to
+    In versus Out:
+    * m.Consume(m.ReadyValid[T])
+    * m.Product(m.ReadyValid[T])
+
+    T is the type of data to be wrapped in Ready/Valid
+
+    The actual semantics of ready/valid are enforced via the use of concrete
+    subclasses.
+    """
+    def fired(self):
+        """
+        Returns a bit that is high when ready and valid are both high
+        """
+        if self.is_consumer():
+            if not self.ready.driven():
+                raise ReadyValidException(
+                    "Cannot invoke fired on consumer when ready isn't driven"
+                )
+
+            return self.valid & self.ready.value()
+
+        assert self.is_producer()
+        if not self.valid.driven():
+            raise ReadyValidException(
+                "Cannot invoke fired on consumer when valid isn't driven"
+            )
+        return self.valid.value() & self.ready
+
+    def enq(self, value, when=True):
+        """
+        Produces the data `value` when `when` is high
+
+        Uses "last connect semantics", which allows multiple `enq` methods to
+        be invoked with different conditions with precedence given to the later
+        invocations.
+        """
+        if not self.is_producer():
+            raise TypeError("Cannot invoke enq on consumer")
+        if when is not True:
+            if not (self.valid.driven() and self.data.driven()):
+                raise ReadyValidException(
+                    "Cannot use enq when without default driver"
+                )
+            curr_valid = self.valid.value()
+            when |= curr_valid
+            self.valid.unwire(curr_valid)
```

Comment:
In this case, we don't want the internal code to be producing a warning.  The ready/valid semantics allow for multiple drivers in the context of conditional assignments, so you could have something like 
```
queue.push(x, when=cond_x)
queue.push(y, when=cond_y)
``

This is valid code where I don't think a warning should be produced, but the internal code needs to unwire the existing driver to do this.  This is partially an artifact of the logic, another option would be to "stage" the conditional drivers and at compile time generate the mux tree for implementing them (right now we build the mux tree as we go, so when we add a conditional push, we unwire the existing driver and add a new mux as the driver).  A third option would be to add a flag to the wire logic that says "ignore_existing_driver" that suppresses the warning when the user is intentionally overwriting a driver.

Reply:
Didn't realize there could be multiple when's. That makes sense.

Since we're already doing `.value()` we could just predicate the `unwire()` call on an `is not None` check on the driver. I think `unwire()` is robust to a `None` value but maybe more clear (and also slightly more performant).

Reply:
Hmm, not exactly sure what you mean by predicate the unwire call, could you clarify?  We want to always do the unwire call, since we assume that there is driver (we raise an Exception if there is no existing driver, because there is no default value, which is why we check driven() first).  

Do you mean have unwire check if the unwire value is None and raise an exception otherwise (rather than do it at this level)?  Right now it will give you an exception trying to remove None from a list, so we could have it raise a specific exception and catch it here and then raise the RVException, but I'm not sure how much that helps, because we're just pushing the None check down into the unwire function, so I don't know if it saves us anything performance or clarity wise.  

I think the only improvement would be if we allowed the default driver to be added later (i.e. we don't raise an exception if the driver is None).  However, I think if we go with this approach, it may be better to go with the staged condition approach I mentioned, since I think we need to delay the creation of the mux tree until we have the default value so we can get the priority right (the other option is to have the default be a "temporary value" that we wire up later, which could also work).

Reply:
I see...valid being driven is checked on L73. *If it wasn't*, I would have suggested doing

```python
if curr_valid is None:
    valid.unwire(curr_valid)
```

but we don't need it because we're checking that before.